kafka-payload-processor/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaPayloadProcessor.java

```diff
@@ -0,0 +1,136 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.concurrent.FastThreadLocal;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Consumer;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.MessagePayload;
+import org.apache.pulsar.client.api.MessagePayloadContext;
+import org.apache.pulsar.client.api.MessagePayloadProcessor;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.impl.MessagePayloadImpl;
+import org.apache.pulsar.client.impl.MessagePayloadUtils;
+import org.apache.pulsar.common.allocator.PulsarByteBufAllocator;
+import org.apache.pulsar.common.api.proto.SingleMessageMetadata;
+
+/**
+ * Process Kafka messages so that Pulsar consumer can recognize.
+ */
+public class KafkaPayloadProcessor implements MessagePayloadProcessor {
+
+    private static final StringDeserializer deserializer = new StringDeserializer();
+    private static final FastThreadLocal<SingleMessageMetadata> LOCAL_SINGLE_MESSAGE_METADATA =
+            new FastThreadLocal<SingleMessageMetadata>() {
+        @Override
+        protected SingleMessageMetadata initialValue() throws Exception {
+            return new SingleMessageMetadata();
+        }
+    };
+
+    @Override
+    public <T> void process(MessagePayload payload,
+                            MessagePayloadContext context,
+                            Schema<T> schema,
+                            Consumer<Message<T>> messageConsumer) throws Exception {
+        if (!isKafkaFormat(context)) {
+            DEFAULT.process(payload, context, schema, messageConsumer);
+            return;
+        }
+
+        final ByteBuf buf = MessagePayloadUtils.convertToByteBuf(payload);
+        try {
+            final MemoryRecords records = MemoryRecords.readableRecords(buf.nioBuffer());
+            int numMessages = 0;
+            for (Record ignored : records.records()) {
+                numMessages++;
+            }
+            int index = 0;
+            for (RecordBatch batch : records.batches()) {
+                if (batch.isControlBatch() || batch.isTransactional()) {
```

Comment:
The normal batch if is sent by transaction, then the `batch.isTransactional` is always true.  We only need to skip when the batch is the control batch.

Reply:
Good point, addressed.